include/seqan3/alignment/matrix/alignment_trace_matrix.hpp

```diff
@@ -0,0 +1,339 @@
+ */
+enum struct trace_matrix_directions : uint8_t
+{
+    //!\brief No trace
+    none      = 0,
+    //!\brief Trace comes from the diagonal entry.
+    diagonal  = 1,
+    //!\brief Trace comes from the above entry.
+    up        = 2,
+    //!\brief Trace comes from the left entry.
+    left      = 4
+};
+
+} // namespace seqan3::detail
+
+namespace seqan3
+{
+//!\brief Enable bitwise operators for enum seqan3::detail::trace_matrix_directions.
+//!\ingroup alignment_matrix
+template <>
+constexpr bool add_enum_bitwise_operators<seqan3::detail::trace_matrix_directions> = true;
+} // namespace seqan3
+
+//!\cond
+namespace seqan3::detail
+{
+template <typename config_type>
+concept has_band = requires(config_type config)
+{
+    {config.lower_band};
+    {config.upper_band};
+};
+}
+//!\endcond
+
+namespace seqan3::detail
+{
+
+//!\brief The declaration of alignment_trace_matrix. Each definition of this
+//!       declaration must satisfy seqan3::detail::matrix_concept.
+//!\ingroup alignment_matrix
+//!\implements seqan3::detail::matrix_concept
+template <typename ...>
+struct alignment_trace_matrix;
+
+/*!\brief Compute the trace from a trace matrix
+ * \ingroup alignment_matrix
+ * \tparam database_t     The type of the database sequence.
+ * \tparam query_t        The type of the query sequence.
+ * \tparam trace_matrix_t The type of the trace matrix.
+ * \param  database       The database sequence.
+ * \param  query          The query sequence.
+ * \param  matrix         The trace matrix.
+ */
+template <
+    typename database_t,
+    typename query_t,
+    typename trace_matrix_t,
+    typename gapped_database_t = gapped<std::decay_t<decltype(std::declval<database_t>()[0])>>,
+    typename gapped_query_t = gapped<std::decay_t<decltype(std::declval<query_t>()[0])>>>
+//!\cond
+    requires matrix_concept<std::remove_reference_t<trace_matrix_t>> &&
+             std::Same<typename std::remove_reference_t<trace_matrix_t>::entry_type, trace_matrix_directions>
+//!\endcond
+inline std::pair<std::list<gapped_database_t>, std::list<gapped_query_t>>
+alignment_trace(database_t && database, query_t && query, trace_matrix_t && matrix)
+{
+    constexpr auto N = trace_matrix_directions::none;
+    constexpr auto D = trace_matrix_directions::diagonal;
+    constexpr auto L = trace_matrix_directions::left;
+    constexpr auto U = trace_matrix_directions::up;
+    int row = matrix.rows()-1;
+    int col = matrix.cols()-1;
+
+    std::list<gapped_database_t> gapped_database{};
+    std::list<gapped_query_t> gapped_query{};
+
+    if (matrix.at(0, 0) != N)
+        throw std::runtime_error{"End trace must be NONE"};
+
+    while(row >= 0 && col >= 0)
+    {
+        trace_matrix_directions dir = matrix.at(row, col);
+        if ((dir & L) == L)
+        {
+            col = std::max<int>(col - 1, 0);
+            gapped_database.push_front(database[col]);
+            gapped_query.push_front(gap::GAP);
```

Comment:
`insert_gap` assumes that I have an underlying sequence already. But in this case I always push a character to the front. Furthermore, I already know my `gapped_sequence` datatype.

I could change that to `insert_gap(gapped_database, std::begin(gapped_database))`, but I don't think this is better than simply calling `push_front`.

Reply:
In general I would construct the a temporary gapped_sequence in reverse order and then reverse move it into the target sequences value by value. It still would be 2n runtime but I think more efficient then constructing lists (not sure if it does small value optimisations on the nodes) and then converting them to gapped sequences.
But we will fix that later. Right now, it is important to have a working code base and that the interfaces can be streamlined.

Reply:
> In general I would construct the a temporary gapped_sequence in reverse order and then reverse move it into the target sequences value by value. It still would be 2n runtime but I think more efficient then constructing lists

absolutely! a std::list of single byte type has an overhead of 16x over the actual size and bad cache behaviour and slower access because locality cannot be assumed.

> I could change that to insert_gap(gapped_database, std::begin(gapped_database)), but I don't think this is better than simply calling push_front.

Well we want to be able to specify other `aligned_sequence_concept` types as return values...